Common++/header/Logger.h

```diff
@@ -284,17 +316,82 @@ namespace pcpp
 			return instance;
 		}
 
+		template <class T> friend void pcpp::log(LogSource source, LogLevel level, T const& message);
+
 	private:
 		bool m_LogsEnabled;
-		Logger::LogLevel m_LogModulesArray[NumOfLogModules];
+		LogLevel m_LogModulesArray[NumOfLogModules];
 		LogPrinter m_LogPrinter;
 		std::string m_LastError;
-		std::ostringstream* m_LogStream;
 
 		// private c'tor - this class is a singleton
 		Logger();
 
+		void printLogMessage(LogSource source, LogLevel logLevel, std::string const& message);
 		static void defaultLogPrinter(LogLevel logLevel, const std::string& logMessage, const std::string& file,
 		                              const std::string& method, const int line);
 	};
+
+	template <class T> inline void log(LogSource source, LogLevel level, T const& message)
+	{
+		auto& logger = Logger::getInstance();
+		if (logger.shouldLog(level, source.logModule))
+		{
+			std::ostringstream sstream;
+			sstream << message;
+			logger.printLogMessage(source, level, sstream.str());
+		}
+	};
+
+	// Specialization for string to skip the stringstream
+	template <> inline void log(LogSource source, LogLevel level, std::string const& message)
+	{
+		auto& logger = Logger::getInstance();
+		if (logger.shouldLog(level, source.logModule))
+		{
+			logger.printLogMessage(source, level, message);
+		}
+	};
+
+	// Specialization for const char* to skip the stringstream
+	template <> inline void log(LogSource source, LogLevel level, const char* const& message)
+	{
+		auto& logger = Logger::getInstance();
+		if (logger.shouldLog(level, source.logModule))
+		{
+			logger.printLogMessage(source, level, message);
+		}
+	};
+
+	template <class T> inline void logError(LogSource source, T const& message)
+	{
+		log(source, LogLevel::Error, message);
+	};
+
+	template <class T> inline void logInfo(LogSource source, T const& message)
+	{
+		log(source, LogLevel::Info, message);
+	};
+
+	template <class T> inline void logDebug(LogSource source, T const& message)
+	{
+		log(source, LogLevel::Debug, message);
+	};
 }  // namespace pcpp
+
+#define PCPP_LOG(level, message)                                                                                       \
+	do                                                                                                                 \
+	{                                                                                                                  \
+		if (pcpp::Logger::getInstance().shouldLog(level, LOG_MODULE))                                                  \
+		{                                                                                                              \
+			std::ostringstream sstream;                                                                                \
+			sstream << message;                                                                                        \
```

Comment:
Can you please check if this change affects the PcapPlusPlus binary sizes?

I made a change in this PR a couple of years ago to significantly reduce the binary sizes: https://github.com/seladb/PcapPlusPlus/pull/748

Reply:
This is a breakdown on the install directory for the x64 release w/npcap.
```
File                                                     Folder1Size Folder2Size AbsoluteDelta RelativeDelta
----                                                     ----------- ----------- ------------- -------------
bin\Arping.exe                                                714240      639488         74752 10.47%
bin\ArpSpoofing.exe                                           703488      631296         72192 10.26%
bin\benchmark.exe                                             599552      562688         36864 6.15%
bin\DisplayLiveDevices.exe                                     30208       30208             0 0.00%
bin\DNSResolver.exe                                           748032      664064         83968 11.23%
bin\DnsSpoofing.exe                                           747520      671232         76288 10.21%
bin\HttpAnalyzer.exe                                          832512      743936         88576 10.64%
bin\IcmpFileTransfer-catcher.exe                              730624      658432         72192 9.88%
bin\IcmpFileTransfer-pitcher.exe                              742400      666112         76288 10.28%
bin\IPDefragUtil.exe                                          711680      655360         56320 7.91%
bin\IPFragUtil.exe                                            693248      637952         55296 7.98%
bin\PcapPrinter.exe                                           658432      612352         46080 7.00%
bin\PcapSearch.exe                                            661504      618496         43008 6.50%
bin\PcapSplitter.exe                                          720384      667648         52736 7.32%
bin\SSLAnalyzer.exe                                           822272      733184         89088 10.83%
bin\TcpReassembly.exe                                         799232      710144         89088 11.15%
bin\TLSFingerprinting.exe                                     796672      712704         83968 10.54%
lib\Common++.lib                                             5045102     4887832        157270 3.12%
lib\Packet++.lib                                            19849740    18454316       1395424 7.03%
lib\Pcap++.lib                                               3620658     2896688        723970 20.00%
```

Full breakdown: [npcap-x64-release.txt](https://github.com/user-attachments/files/17369083/npcap-x64-release.txt)

Reply:
Did you test it with MSVC or MinGW? Also, did you have a chance to test on Linux also?

I don't remember in which platform the impact was the highest, but I think it was Linux 🤔

Reply:
Msvc, I dont rhink I have mingw setup. I should be able to test the sizes on Unix via WSL.

Reply:
@Dimi1010 I see you're still working on it. Please let me know when the changes are done and when you want me to review the PR again.
It'd be good to make sure binary sizes don't increase. I'd test at least on Windows and Linux

Reply:
@seladb Ok, ended up keeping the optimization.

As bc82caf: Release builds ended up being:
- Windows
  - Static Libraries
    - Common++.lib - 4.7MB -> 5MB
    - Packet++.lib - 17.6MB -> 19.7MB
    - Pcap++.lib - 2.8MB -> 3.3MB
  - Examples
    - Same binary size or <=10KB increase.
- Linux (WSL)
  - Static Libraries
    - libCommon++.a - 408KB -> 444KB
    - libPakcet++.a - 3.1MB -> 3.3MB
    - libPcap++.a - 676KB -> 748KB
  - Examples 
    - On average ~0.1MB increase on each binary.